contrib/server/operations/pbench-upload-results.py

```diff
@@ -201,23 +208,29 @@ def main() -> int:
             if str(t) in processed:
                 skipped += 1
                 if parsed.verify:
-                    print(f"[CPT] skip {t}")
+                    print(f"[checkpoint] skip {t}")
                 continue
             which.append(t)
     elif parsed.tarball.is_file():
         which = [parsed.tarball]
     else:
-        print(f"Path {parsed.tarball} doesn't exist", file=sys.stderr)
+        print(f"Path {parsed.tarball} isn't a directory or file", file=sys.stderr)
         return 1
     if parsed.verify:
         print(
             f"Identified {len(which)} target tarballs: skipped {skipped}, {early} too old, {late} too new"
         )
 
+    # We'll append successful uploads to the checkpoint file.
     checkwriter: Optional[TextIOWrapper] = None
     if checkpoint:
         checkwriter = checkpoint.open(mode="a")
 
+    # Now start the upload, checkpointing each file after a successful upload
+    #
+    # We maintain the checkpoint file for --preview to assist in testing: if
+    # you want to test with --preview and then do a real upload, delete the
+    # checkpoint file!
```

Comment:
More to the point, if you are _recovering_ from a failure, **make a copy** of your checkpoint file before trying the preview!!

(If attempting to do a real upload after a preview, it will quickly become obvious what has happened; however, without having created a copy of the checkpoint file, it might not be obvious what to do next....)

Reply:
Arguably. Basically, it's a special purpose admin tool, and you have to know what you're doing and be moderately careful. `--preview` was really useful for debugging, but I don't think I'd be likely to bother for real.

Also, the script *does* gracefully handle duplicates, so the checkpoint is a convenience (and potentially a massive timesaver), but if you just retry the same command after a failure it's not going to be messed up by the fact that some tarballs were already uploaded ... it'll just take longer.